File: src/minimcp/responder.py
import json
import logging
from operator import attrgetter
from typing import Final

import mcp.types as types

from minimcp.limiter import TimeLimiter
from minimcp.types import Message, Send
from minimcp.utils import json_rpc

logger = logging.getLogger(__name__)

_UNSET: Final[str] = "UNSET"


class Responder:
    """
    Responder enables message handlers to send notifications back to the client.

    The Responder is available in the handler context via mcp.context.get_responder()
    and provides methods for sending server-to-client notifications, including
    progress updates during long-running operations.

    The responder automatically resets the idle timeout when sending notifications,
    ensuring the handler doesn't time out while actively communicating with the client.
    """

    _request: Message
    _send: Send
    _time_limiter: TimeLimiter
    _progress_token: types.ProgressToken | None

    def __init__(self, request: Message, send: Send, time_limiter: TimeLimiter):
        """
        Args:
            request: The incoming message that triggered the handler.
            send: The send function for transmitting messages to the client.
            time_limiter: The TimeLimiter for managing handler idle timeout.
        """
        self._request = request
        self._send = send
        self._time_limiter = time_limiter
        self._progress_token = _UNSET

    def _get_progress_token(self) -> types.ProgressToken | None:
        """Extract and cache the progress token from the request metadata.

        Parsing is deferred to the first call so that handlers that never invoke
        ``report_progress`` pay no deserialization cost.

        Returns:
            The progress token provided by the client in the request metadata,
            or None if no token was supplied or the request could not be parsed.
        """
        if self._progress_token is _UNSET:
            try:
                client_request = types.ClientRequest.model_validate(json.loads(self._request))
                self._progress_token = attrgetter("params.meta.progressToken")(client_request.root)
            except Exception:
                self._progress_token = None
        return self._progress_token

    async def report_progress(
        self, progress: float, total: float | None = None, message: str | None = None
    ) -> types.ProgressToken | None:
        """Report progress for the current operation to the client.

        This method sends a progress notification to the client, useful for long-running
        operations where you want to keep the client informed. The notification is only
        sent if the client provided a progress token in the request metadata.

        The idle timeout is automatically reset when sending the progress notification.

        Args:
            progress: Current progress value (e.g., 24 for 24 items processed).
            total: Optional total value (e.g., 100 for total items). Useful for
                calculating completion percentage.
            message: Optional human-readable message to display to the user
                (e.g., "Processing file 24 of 100").

        Returns:
            The progress token if the notification was sent successfully, None otherwise.
            Returns None if no progress token was provided in the request.
        """

        progress_token = self._get_progress_token()

        if progress_token is None:
            logger.warning("report_progress failed: Progress token is not available.")
            return None

        notification = types.ServerNotification(
            types.ProgressNotification(
                method="notifications/progress",  # TODO: Remove once python-sdk/pull/1292 is merged.
                params=types.ProgressNotificationParams(
                    progressToken=progress_token,
                    progress=progress,
                    total=total,
                    message=message,
                ),
            )
        )

        await self.send_notification(notification)

        return progress_token

    async def send_notification(self, notification: types.ServerNotification) -> None:
        """Send a notification to the client.

        Notifications are one-way messages from the server to the client that do not
        expect a response. This enables bidirectional communication in MiniMCP servers,
        allowing handlers to proactively push updates to the client.

        The idle timeout is automatically reset when sending notifications, ensuring
        the handler remains active while communicating with the client.

        Args:
            notification: The server notification to send. Common notification types
                include progress notifications, log messages, and resource updates etc.
        """
        logger.debug("Sending notification: %s", notification)
        message = json_rpc.build_notification_message(notification)

        # Reset time limiter
        self._time_limiter.reset()

        # Just call the sender with the message and let transport layer handle the rest.
        await self._send(message)

File: rust/crates/minimcp/src/responder.rs
//! Server-to-client notifications.
//!
//! Mirrors `minimcp/responder.py`. A `Responder` is available to handlers (via
//! the request context) on transports that support bidirectional communication
//! and lets handlers push notifications, such as progress updates, to the
//! client.

use serde_json::{json, Value};

use crate::mcp_types::JSONRPC_VERSION;
use crate::time_limiter::TimeLimiter;
use crate::types::SendFn;

/// Enables message handlers to send notifications back to the client.
///
/// Sending a notification resets the idle timeout so an actively communicating
/// handler is not timed out.
#[derive(Clone)]
pub struct Responder {
    send: SendFn,
    time_limiter: Option<TimeLimiter>,
}

impl Responder {
    /// Create a responder.
    ///
    /// # Arguments
    ///
    /// * `send` - The callback used to transmit messages to the client.
    /// * `time_limiter` - The handler's idle-timeout limiter, if enabled.
    pub fn new(send: SendFn, time_limiter: Option<TimeLimiter>) -> Self {
        Self { send, time_limiter }
    }

    /// Send a JSON-RPC notification with the given method and params.
    ///
    /// The idle timeout is reset before the message is sent.
    pub async fn send_notification(&self, method: &str, params: Value) {
        if let Some(limiter) = &self.time_limiter {
            limiter.reset();
        }

        let message = json!({
            "jsonrpc": JSONRPC_VERSION,
            "method": method,
            "params": params,
        })
        .to_string();

        (self.send)(message).await;
    }

    /// Report progress for the current operation to the client.
    ///
    /// # Arguments
    ///
    /// * `progress_token` - The token supplied by the client in the request metadata.
    /// * `progress` - Current progress value.
    /// * `total` - Optional total value for computing completion percentage.
    /// * `message` - Optional human-readable status message.
    pub async fn report_progress(
        &self,
        progress_token: Value,
        progress: f64,
        total: Option<f64>,
        message: Option<String>,
    ) {
        let mut params = json!({
            "progressToken": progress_token,
            "progress": progress,
        });
        if let Some(total) = total {
            params["total"] = json!(total);
        }
        if let Some(message) = message {
            params["message"] = json!(message);
        }

        self.send_notification("notifications/progress", params)
            .await;
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::sync::{Arc, Mutex};

    fn recording_responder() -> (Responder, Arc<Mutex<Vec<String>>>) {
        let recorded: Arc<Mutex<Vec<String>>> = Arc::new(Mutex::new(Vec::new()));
        let sink = recorded.clone();
        let send: SendFn = Arc::new(move |msg: String| {
            let sink = sink.clone();
            Box::pin(async move {
                sink.lock().unwrap().push(msg);
            })
        });
        let limiter = TimeLimiter::new(std::time::Duration::from_secs(30));
        (Responder::new(send, Some(limiter)), recorded)
    }

    #[tokio::test]
    async fn send_notification_builds_jsonrpc() {
        let (responder, recorded) = recording_responder();
        responder
            .send_notification("notifications/test", json!({"k": "v"}))
            .await;

        let messages = recorded.lock().unwrap();
        assert_eq!(messages.len(), 1);
        let parsed: Value = serde_json::from_str(&messages[0]).unwrap();
        assert_eq!(parsed["jsonrpc"], "2.0");
        assert_eq!(parsed["method"], "notifications/test");
        assert_eq!(parsed["params"]["k"], "v");
        assert!(parsed.get("id").is_none());
    }

    #[tokio::test]
    async fn report_progress_full_params() {
        let (responder, recorded) = recording_responder();
        responder
            .report_progress(
                json!("tok-1"),
                42.5,
                Some(100.0),
                Some("halfway".to_string()),
            )
            .await;

        let messages = recorded.lock().unwrap();
        let parsed: Value = serde_json::from_str(&messages[0]).unwrap();
        assert_eq!(parsed["method"], "notifications/progress");
        assert_eq!(parsed["params"]["progressToken"], "tok-1");
        assert_eq!(parsed["params"]["progress"], 42.5);
        assert_eq!(parsed["params"]["total"], 100.0);
        assert_eq!(parsed["params"]["message"], "halfway");
    }

    #[tokio::test]
    async fn report_progress_omits_optional_fields() {
        let (responder, recorded) = recording_responder();
        responder
            .report_progress(json!("tok-2"), 10.0, None, None)
            .await;

        let messages = recorded.lock().unwrap();
        let parsed: Value = serde_json::from_str(&messages[0]).unwrap();
        assert_eq!(parsed["params"]["progress"], 10.0);
        assert!(parsed["params"].get("total").is_none());
        assert!(parsed["params"].get("message").is_none());
    }

    #[tokio::test]
    async fn multiple_notifications_are_all_sent() {
        let (responder, recorded) = recording_responder();
        for i in 0..5 {
            responder
                .report_progress(json!("tok"), i as f64 * 10.0, None, None)
                .await;
        }
        assert_eq!(recorded.lock().unwrap().len(), 5);
    }
}
